[clang][Sema] crash on invalid structured-binding-like function definition

# Environment

```
clang-18
```
or 
```
clang-23
```

# diagnostic msg

```sh
# Crash reproducer for clang version 23.0.0git (https://github.com/llvm/llvm-project.git aff5afc48df63615053b2432da198a4932435c3f)
# Driver args: "-fsyntax-only" "-std=c++20" "test1.cxx"
# Original command:  "/home/ubuntu2404/llvm-project/build/bin/clang-23" "-cc1" "-triple" "x86_64-unknown-linux-gnu" "-fsyntax-only" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "test1.cxx" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-mframe-pointer=all" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "x86-64" "-tune-cpu" "generic" "-debugger-tuning=gdb" "-fdebug-compilation-dir=/home/ubuntu2404/lsp_fuzz_latest/vscode_poc" "-fcoverage-compilation-dir=/home/ubuntu2404/lsp_fuzz_latest/vscode_poc" "-resource-dir" "/home/ubuntu2404/llvm-project/build/lib/clang/23" "-internal-isystem" "/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13" "-internal-isystem" "/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/x86_64-linux-gnu/c++/13" "-internal-isystem" "/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/backward" "-internal-isystem" "/home/ubuntu2404/llvm-project/build/lib/clang/23/include" "-internal-isystem" "/usr/local/include" "-internal-isystem" "/usr/lib/gcc/x86_64-linux-gnu/13/../../../../x86_64-linux-gnu/include" "-internal-externc-isystem" "/usr/include/x86_64-linux-gnu" "-internal-externc-isystem" "/include" "-internal-externc-isystem" "/usr/include" "-std=c++20" "-fdeprecated-macro" "-ferror-limit" "19" "-fmessage-length=193" "-fgnuc-version=4.2.1" "-fno-implicit-modules" "-fskip-odr-check-in-gmf" "-fcxx-exceptions" "-fexceptions" "-fcolor-diagnostics" "-faddrsig" "-fdwarf2-cfi-asm" "-x" "c++" "test1.cxx"
 "/home/ubuntu2404/llvm-project/build/bin/clang-23" "-cc1" "-triple" "x86_64-unknown-linux-gnu" "-fsyntax-only" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "test1.cxx" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-mframe-pointer=all" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "x86-64" "-tune-cpu" "generic" "-debugger-tuning=gdb" "-fdebug-compilation-dir=/home/ubuntu2404/lsp_fuzz_latest/vscode_poc" "-fcoverage-compilation-dir=/home/ubuntu2404/lsp_fuzz_latest/vscode_poc" "-std=c++20" "-fdeprecated-macro" "-ferror-limit" "19" "-fmessage-length=193" "-fgnuc-version=4.2.1" "-fno-implicit-modules" "-fskip-odr-check-in-gmf" "-fcxx-exceptions" "-fexceptions" "-fcolor-diagnostics" "-faddrsig" "-fdwarf2-cfi-asm" "-x" "c++" "test1-d83ba9.cpp"
```

```cpp
# 1 "test1.cxx"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 493 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "test1.cxx" 2
([a])(){a}

```


# Reproducer
```cpp
([a])(){a}
```

# Command
```
clang -fsyntax-only -std=c++20 test1.cxx
```
# Stack trace

```
test1.cxx:1:2: error: a type specifier is required for all declarations
    1 | ([a])(){a}
      |  ^
test1.cxx:1:2: error: decomposition declaration cannot be declared with parentheses
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: clang -fsyntax-only -std=c++20 test1.cxx
1.      test1.cxx:1:9: current parser token 'a'
2.      test1.cxx:1:8: parsing function body '[a]'
3.      test1.cxx:1:8: in compound statement ('{}')
 #0 0x000074bd985a63bf llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0xda63bf)
 #1 0x000074bd985a44f9 llvm::sys::RunSignalHandlers() (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0xda44f9)
 #2 0x000074bd984f0227 (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0xcf0227)
 #3 0x000074bd97045330 (/lib/x86_64-linux-gnu/libc.so.6+0x45330)
 #4 0x000074bd9fe2f6d7 clang::DeclContext::lookup(clang::DeclarationName) const (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xe2f6d7)
 #5 0x000074bda08d8605 (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x18d8605)
 #6 0x000074bda08d577f clang::Sema::LookupQualifiedName(clang::LookupResult&, clang::DeclContext*, bool) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x18d577f)
 #7 0x000074bda08d4451 clang::Sema::CppLookupName(clang::LookupResult&, clang::Scope*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x18d4451)
 #8 0x000074bda08d7ee1 clang::Sema::LookupName(clang::LookupResult&, clang::Scope*, bool, bool) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x18d7ee1)
 #9 0x000074bda0565894 clang::Sema::ClassifyName(clang::Scope*, clang::CXXScopeSpec&, clang::IdentifierInfo*&, clang::SourceLocation, clang::Token const&, clang::CorrectionCandidateCallback*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x1565894)
#10 0x000074bd9fc65415 clang::Parser::TryAnnotateName(clang::CorrectionCandidateCallback*, clang::ImplicitTypenameContext) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc65415)
#11 0x000074bd9fc3f259 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributes&, clang::ParsedAttributes&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc3f259)
#12 0x000074bd9fc3edfc clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc3edfc)
#13 0x000074bd9fc47aa2 clang::Parser::ParseCompoundStatementBody(bool) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc47aa2)
#14 0x000074bd9fc48b77 clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc48b77)
#15 0x000074bd9fc640c7 clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc640c7)
#16 0x000074bd9fbad2a1 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::ParsedAttributes&, clang::SourceLocation*, clang::Parser::ForRangeInit*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xbad2a1)
#17 0x000074bd9fc6307f clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc6307f)
#18 0x000074bd9fc62a39 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc62a39)
#19 0x000074bd9fc61c65 clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc61c65)
#20 0x000074bd9fc5ff7b clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc5ff7b)
#21 0x000074bd9fc5f8b6 clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xc5f8b6)
#22 0x000074bd9fb97370 clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xb97370)
#23 0x000074bda1a0662c clang::FrontendAction::Execute() (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x2a0662c)
#24 0x000074bda19830b4 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x29830b4)
#25 0x000074bda1a8263d clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x2a8263d)
#26 0x00006115a067242e cc1_main(llvm::ArrayRef<char const*>, char const*, void*) build-llvm/tools/clang/stage2-bins/clang/tools/driver/cc1_main.cpp:294:15
#27 0x00006115a066f894 ExecuteCC1Tool build-llvm/tools/clang/stage2-bins/clang/tools/driver/driver.cpp:365:12
#28 0x000074bda1633972 (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x2633972)
#29 0x000074bd984eff77 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0xceff77)
#30 0x000074bda1633237 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x2633237)
#31 0x000074bda15fb518 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x25fb518)
#32 0x000074bda15fb77f clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x25fb77f)
#33 0x000074bda1617c20 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x2617c20)
#34 0x00006115a066f1ec clang_main(int, char**, llvm::ToolContext const&) build-llvm/tools/clang/stage2-bins/clang/tools/driver/driver.cpp:540:21
#35 0x00006115a067c383 main build-llvm/tools/clang/stage2-bins/build-llvm/tools/clang/stage2-bins/tools/clang/tools/driver/clang-driver.cpp:17:10
#36 0x000074bd9702a1ca __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#37 0x000074bd9702a28b call_init ./csu/../csu/libc-start.c:128:20
#38 0x000074bd9702a28b __libc_start_main ./csu/../csu/libc-start.c:347:5
#39 0x00006115a066c255 _start (/usr/lib/llvm-18/bin/clang+0xe255)
clang: error: clang frontend command failed with exit code 139 (use -v to see invocation)
Ubuntu clang version 18.1.3 (1ubuntu1)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/test1-cc4c9d.cpp
clang: note: diagnostic msg: /tmp/test1-cc4c9d.sh
clang: note: diagnostic msg: 
```



```
test1.cxx:1:2: error: a type specifier is required for all declarations
    1 | ([a])(){a}
      |  ^
test1.cxx:1:2: error: structured binding declaration cannot be declared with parentheses
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /home/ubuntu2404/llvm-project/build_asan/bin/clang -fsyntax-only -std=c++20 test1.cxx
1.      test1.cxx:1:8: current parser token '{'
 #0 0x000062607425a138 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/ubuntu2404/llvm-project/llvm/lib/Support/Unix/Signals.inc:842:13
 #1 0x0000626074258023 llvm::sys::RunSignalHandlers() /home/ubuntu2404/llvm-project/llvm/lib/Support/Signals.cpp:109:18
 #2 0x00006260741d3119 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) /home/ubuntu2404/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:73:5
 #3 0x00006260741d3119 CrashRecoverySignalHandler(int) /home/ubuntu2404/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:390:51
 #4 0x000071c203c45330 (/lib/x86_64-linux-gnu/libc.so.6+0x45330)
 #5 0x0000626076686337 llvm::IntrusiveRefCntPtr<clang::ExternalASTSource>::get() const /home/ubuntu2404/llvm-project/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h:205:27
 #6 0x0000626076686337 clang::ASTContext::getExternalSource() const /home/ubuntu2404/llvm-project/clang/include/clang/AST/ASTContext.h:1390:27
 #7 0x0000626076686337 clang::DeclContext::LoadLexicalDeclsFromExternalStorage() const /home/ubuntu2404/llvm-project/clang/lib/AST/DeclBase.cpp:1609:53
 #8 0x0000626076686af2 clang::DeclContext::decls_begin() const /home/ubuntu2404/llvm-project/clang/lib/AST/DeclBase.cpp:1675:24
 #9 0x0000626075d5141b clang::operator!=(clang::DeclContext::decl_iterator, clang::DeclContext::decl_iterator) /home/ubuntu2404/llvm-project/clang/include/clang/AST/DeclBase.h:2365:24
#10 0x0000626075d5141b clang::Sema::ActOnStartOfFunctionDef(clang::Scope*, clang::Decl*, clang::SkipBodyInfo*, clang::Sema::FnBodyKind) /home/ubuntu2404/llvm-project/clang/lib/Sema/SemaDecl.cpp:16267:20
#11 0x0000626075d50bf6 clang::Sema::ActOnStartOfFunctionDef(clang::Scope*, clang::Declarator&, llvm::MutableArrayRef<clang::TemplateParameterList*>, clang::SkipBodyInfo*, clang::Sema::FnBodyKind) /home/ubuntu2404/llvm-project/clang/lib/Sema/SemaDecl.cpp:15923:15
#12 0x0000626075ae03ce clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) /home/ubuntu2404/llvm-project/clang/lib/Parse/Parser.cpp:1332:23
#13 0x0000626075afda25 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::ParsedAttributes&, clang::Parser::ParsedTemplateInfo&, clang::SourceLocation*, clang::Parser::ForRangeInit*) /home/ubuntu2404/llvm-project/clang/lib/Parse/ParseDecl.cpp:0:0
#14 0x0000626075adf87f clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec&, clang::AccessSpecifier) /home/ubuntu2404/llvm-project/clang/lib/Parse/Parser.cpp:0:10
#15 0x0000626075adf303 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier) /home/ubuntu2404/llvm-project/clang/lib/Parse/Parser.cpp:1170:12
#16 0x0000626075ade6d0 clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) /home/ubuntu2404/llvm-project/clang/lib/Parse/Parser.cpp:0:14
#17 0x0000626075adcdb7 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) /home/ubuntu2404/llvm-project/clang/lib/Parse/Parser.cpp:706:10
#18 0x0000626075adc9b1 clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) /home/ubuntu2404/llvm-project/clang/lib/Parse/Parser.cpp:602:23
#19 0x0000626075ad23a8 clang::ParseAST(clang::Sema&, bool, bool) /home/ubuntu2404/llvm-project/clang/lib/Parse/ParseAST.cpp:169:5
#20 0x0000626074ab4cdd clang::FrontendAction::Execute() /home/ubuntu2404/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1326:10
#21 0x0000626074a2f5e4 llvm::Error::getPtr() const /home/ubuntu2404/llvm-project/llvm/include/llvm/Support/Error.h:281:12
#22 0x0000626074a2f5e4 llvm::Error::operator bool() /home/ubuntu2404/llvm-project/llvm/include/llvm/Support/Error.h:241:16
#23 0x0000626074a2f5e4 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/ubuntu2404/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:1007:23
#24 0x0000626074b7ce09 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/ubuntu2404/llvm-project/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:310:25
#25 0x0000626072051266 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/ubuntu2404/llvm-project/clang/tools/driver/cc1_main.cpp:304:15
#26 0x000062607204df1c ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>) /home/ubuntu2404/llvm-project/clang/tools/driver/driver.cpp:226:12
#27 0x000062607204f6cd std::__atomic_base<int>::fetch_sub(int, std::memory_order) /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/atomic_base.h:645:16
#28 0x000062607204f6cd llvm::ThreadSafeRefCountedBase<llvm::vfs::FileSystem>::Release() const /home/ubuntu2404/llvm-project/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h:134:32
#29 0x000062607204f6cd llvm::IntrusiveRefCntPtrInfo<llvm::vfs::FileSystem>::release(llvm::vfs::FileSystem*) /home/ubuntu2404/llvm-project/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h:164:38
#30 0x000062607204f6cd llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>::release() /home/ubuntu2404/llvm-project/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h:233:7
#31 0x000062607204f6cd llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>::~IntrusiveRefCntPtr() /home/ubuntu2404/llvm-project/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h:196:27
#32 0x000062607204f6cd clang_main(int, char**, llvm::ToolContext const&)::$_0::operator()(llvm::SmallVectorImpl<char const*>&) const /home/ubuntu2404/llvm-project/clang/tools/driver/driver.cpp:376:5
#33 0x000062607204f6cd int llvm::function_ref<int (llvm::SmallVectorImpl<char const*>&)>::callback_fn<clang_main(int, char**, llvm::ToolContext const&)::$_0>(long, llvm::SmallVectorImpl<char const*>&) /home/ubuntu2404/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:46:12
#34 0x00006260748d1e59 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::$_0::operator()() const /home/ubuntu2404/llvm-project/clang/lib/Driver/Job.cpp:442:30
#35 0x00006260748d1e59 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::$_0>(long) /home/ubuntu2404/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:46:12
#36 0x00006260741d2ecc llvm::function_ref<void ()>::operator()() const /home/ubuntu2404/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:0:12
#37 0x00006260741d2ecc llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) /home/ubuntu2404/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:426:3
#38 0x00006260748d1756 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const /home/ubuntu2404/llvm-project/clang/lib/Driver/Job.cpp:442:7
#39 0x000062607489d9bb clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const /home/ubuntu2404/llvm-project/clang/lib/Driver/Compilation.cpp:196:15
#40 0x000062607489db4e clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const /home/ubuntu2404/llvm-project/clang/lib/Driver/Compilation.cpp:246:13
#41 0x00006260748b629d llvm::SmallVectorBase<unsigned int>::empty() const /home/ubuntu2404/llvm-project/llvm/include/llvm/ADT/SmallVector.h:83:46
#42 0x00006260748b629d clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) /home/ubuntu2404/llvm-project/clang/lib/Driver/Driver.cpp:2281:23
#43 0x000062607204da6c clang_main(int, char**, llvm::ToolContext const&) /home/ubuntu2404/llvm-project/clang/tools/driver/driver.cpp:414:21
#44 0x000062607205ae97 main /home/ubuntu2404/llvm-project/build/tools/clang/tools/driver/clang-driver.cpp:17:10
#45 0x000071c203c2a1ca __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#46 0x000071c203c2a28b call_init ./csu/../csu/libc-start.c:128:20
#47 0x000071c203c2a28b __libc_start_main ./csu/../csu/libc-start.c:347:5
#48 0x000062607204bef5 _start (/home/ubuntu2404/llvm-project/build_asan/bin/clang+0xed2ef5)
clang: error: clang frontend command failed with exit code 139 (use -v to see invocation)
clang version 23.0.0git (https://github.com/llvm/llvm-project.git aff5afc48df63615053b2432da198a4932435c3f)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/ubuntu2404/llvm-project/build/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/test1-d83ba9.cpp
clang: note: diagnostic msg: /tmp/test1-d83ba9.sh
clang: note: diagnostic msg: 
```


This crash is found due to reduction towards https://github.com/llvm/llvm-project/issues/184820

--- a/clang/docs/ReleaseNotes.md
+++ b/clang/docs/ReleaseNotes.md
@@ -496,6 +496,10 @@ features cannot lower the translation-unit ABI level;
   parsed where a nested-name-specifier could appear (e.g. ``int decltype = 0;``).
   Clang now diagnoses the error instead of asserting. (#GH211207)
 
+- Fixed an assertion failure when a parenthesized structured binding declarator
+  was followed by a function declarator and body (e.g. ``([a, b])() {}``).
+  (#GH218144, #GH193687)
+
 - Fixed a crash when computing the implicit deletion of a defaulted comparison
   operator required an access check that ran while an enclosing declaration
   was still being parsed. (#GH210692)
@@ -535,10 +539,6 @@ features cannot lower the translation-unit ABI level;
   definition of a member of a class template added a default argument to a
   parameter that follows a parameter pack (e.g.
   `template <typename... T> S::S(T..., int = 10) {}`).  (#GH216211)
-
-- Fixed an assertion failure when a parenthesized structured binding declarator
-  was followed by a function declarator and body (e.g. ``([a, b])() {}``).
-  (#GH218144, #GH193687)
 
 #### Bug Fixes to AST Handling
 

--- a/clang/docs/ReleaseNotes.md
+++ b/clang/docs/ReleaseNotes.md
@@ -536,6 +536,10 @@ features cannot lower the translation-unit ABI level;
   parameter that follows a parameter pack (e.g.
   `template <typename... T> S::S(T..., int = 10) {}`).  (#GH216211)
 
+- Fixed an assertion failure when a parenthesized structured binding declarator
+  was followed by a function declarator and body (e.g. ``([a, b])() {}``).
+  (#GH218144, #GH193687)
+
 #### Bug Fixes to AST Handling
 
 - Fixed a non-deterministic ordering of unused local typedefs that made

--- a/clang/lib/Parse/ParseDecl.cpp
+++ b/clang/lib/Parse/ParseDecl.cpp
@@ -6862,6 +6862,11 @@ void Parser::ParseDirectDeclarator(Declarator &D) {
     // Example: 'char (*X)'   or 'int (*XX)(void)'
     ParseParenDeclarator(D);
 
+    // As above, a structured binding declarator cannot be followed by any
+    // declarator chunks.
+    if (D.isDecompositionDeclarator())
+      return;
+
     // If the declarator was parenthesized, we entered the declarator
     // scope when parsing the parenthesized declarator, then exited
     // the scope already. Re-enter the scope, if we need to.

--- a/clang/test/Parser/GH218144.cpp
+++ b/clang/test/Parser/GH218144.cpp
@@ -1,0 +1,30 @@
+// RUN: %clang_cc1 -std=c++17 -fsyntax-only -verify %s
+// RUN: %clang_cc1 -std=c++26 -fsyntax-only -verify %s
+
+b;([c,ac])(){}
+// expected-error@-1 2 {{a type specifier is required for all declarations}}
+// expected-error@-2 {{structured binding declaration cannot be declared with parentheses}}
+// expected-error@-3 {{expected expression}}
+// expected-error@-4 {{expected ';' after top level declarator}}
+
+([d])(){d}
+// expected-error@-1 {{a type specifier is required for all declarations}}
+// expected-error@-2 {{structured binding declaration cannot be declared with parentheses}}
+// expected-error@-3 {{expected expression}}
+// expected-error@-4 {{expected ';' after top level declarator}}
+
+auto ([a])() {}
+// expected-error@-1 {{structured binding declaration cannot be declared with parentheses}}
+// expected-error@-2 {{expected expression}}
+// expected-error@-3 {{expected ';' after top level declarator}}
+
+auto (([x, y, z]))() {}
+// expected-error@-1 {{structured binding declaration cannot be declared with parentheses}}
+// expected-error@-2 {{expected expression}}
+// expected-error@-3 {{expected ';' after top level declarator}}
+
+struct S {
+  ([m, n])() {}
+  // expected-error@-1 {{structured binding declaration not permitted in this context}}
+  // expected-error@-2 {{expected ';' at end of declaration list}}
+};

